File: frontend/src/services/admin-service.ts
import { apiClient } from "@/lib/api-client";
import { unwrap } from "@/lib/api-helpers";
import type { ApiResponse, PagedResponse, PageParams } from "@/types/api";
import type {
  ApprovalStatus,
  CompanyResponse,
  CompanySummaryResponse,
  RejectCompanyRequest,
} from "@/types/domain";

/** Admin company-review endpoints under /api/v1/admin/companies. */
export const adminService = {
  async listCompanies(
    status: ApprovalStatus = "PENDING",
    params: PageParams = {},
  ): Promise<PagedResponse<CompanySummaryResponse>> {
    const { data } = await apiClient.get<ApiResponse<PagedResponse<CompanySummaryResponse>>>(
      "/admin/companies",
      { params: { status, ...params } },
    );
    return unwrap(data);
  },

  async getCompany(companyId: string): Promise<CompanyResponse> {
    const { data } = await apiClient.get<ApiResponse<CompanyResponse>>(
      `/admin/companies/${companyId}`,
    );
    return unwrap(data);
  },

  async approve(companyId: string): Promise<CompanyResponse> {
    const { data } = await apiClient.post<ApiResponse<CompanyResponse>>(
      `/admin/companies/${companyId}/approve`,
    );
    return unwrap(data);
  },

  async reject(companyId: string, payload: RejectCompanyRequest): Promise<CompanyResponse> {
    const { data } = await apiClient.post<ApiResponse<CompanyResponse>>(
      `/admin/companies/${companyId}/reject`,
      payload,
    );
    return unwrap(data);
  },
};

File: src/main/java/com/campusconnect/portal/service/admin/AdminService.java
package com.campusconnect.portal.service.admin;

import com.campusconnect.portal.common.enums.ApprovalStatus;
import com.campusconnect.portal.dto.admin.CompanyResponse;
import com.campusconnect.portal.dto.admin.CompanySummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

/**
 * Platform-administration operations. Restricted to the {@code ADMIN} role at the controller
 * layer. Currently scoped to company onboarding: reviewing, approving, and rejecting the
 * companies that register through {@code /auth/register/company} in {@code PENDING} status.
 */
public interface AdminService {

    /** Paginated companies filtered by approval status. */
    Page<CompanySummaryResponse> listCompanies(ApprovalStatus status, Pageable pageable);

    /** Full detail for a single company. */
    CompanyResponse getCompany(UUID companyId);

    /**
     * Approves a company so it can post jobs. No-op-safe on an already-approved company;
     * rejects the transition from a terminal state otherwise.
     */
    CompanyResponse approveCompany(UUID companyId);

    /** Rejects a pending company with a reason surfaced to the company. */
    CompanyResponse rejectCompany(UUID companyId, String reason);
}
